transcom/mymove
# [MB-1575] Implementing payment request lookup for zip5 distance

## Description

This PR adds the functionality to look up the zip5 distance when a payment request is made.

## Reviewer Notes

Because the lookup needs to access the route planner, I changed some of the calling code to pass the planner object through.

This is currently checking the distance between the pickup and destination zip5s, not the specific addresses in those zip5s.  Should it be using the addresses instead?

Note that pickup address and destination address are currently nullable fields in the database and the model.  However, the model validations enforce that the addresses are present.  I do a safety check to make sure they are available, but there's some additional work happening to sort out the inconsistency.  See this [Slack thread](https://ustcdp3.slack.com/archives/CP497TGAU/p1588607779167500).

It is possible to have a secondary pickup and/or destination address.  There is setup work that still needs to be done to add those as part of the calculations.  This has been captured in [another story](https://dp3.atlassian.net/browse/MB-2526).

## Setup

### Testing Dev

1. First run `make db_dev_e2e_populate` to ensure your database is setup with the E2E data.
1. Next, save this example JSON payload (which references E2E UUIDs) to a file called `pr_payload_dev.json`:
    ```json
    {
      "isFinal": false,
      "moveTaskOrderID": "da3f34cc-fb94-4e0b-1c90-ba3333cb7791",
      "serviceItems": [
        {
          "id": "3624d82f-fa87-47f5-a09a-2d5639e45c02"
        }
      ]
    }
    ```
    Note that the service item ID above is referencing the service `DLH` (domestic linehaul), which has `RequestedPickupDate` as an expected parameter.
1. Ensure your server is running: `make server_run`
1. Run this command to fetch the "available to prime" MTOs and look at the payment request state for the `moveTaskOrderID` above in the returned JSON: `go run ./cmd/prime-api-client --insecure fetch-mtos | jq`
1. Look at the `payment_service_item_params` table and you should not see any records.
1. Now let's create a payment request using our payload above.  There are two different ways to do this:
    - With the `--filename` flag: `go run ./cmd/prime-api-client --insecure create-payment-request --filename pr_payload_dev.json | jq`
    - From stdin: `go run ./cmd/prime-api-client --insecure create-payment-request - < pr_payload_dev.json | jq`
1. Fetch the MTOs again and verify the payment request(s) you created are there: `go run ./cmd/prime-api-client --insecure fetch-mtos | jq`
1. Look at the `payment_service_item_params` table and verify that there are now params that include `DistanceZip5 ` and that the value is the appropriate mileage from between the addresses in `mto_shipment`.
1. This query will also show you the key/value param pairs for an MTO service item id:
    ```sql
    select key, value from payment_service_item_params
    inner join payment_service_items psi on payment_service_item_params.payment_service_item_id = psi.id
    inner join service_item_param_keys sipk on payment_service_item_params.service_item_param_key_id = sipk.id
    where mto_service_item_id = '3624d82f-fa87-47f5-a09a-2d5639e45c02';
    ```

## Code Review Verification Steps


## References

* [Jira story](https://dp3.atlassian.net/browse/MB-1575) for this change
